### PR TITLE
feat: support all model providers in JSON agent configuration

### LINKED ISSUE #1064: [FEATURE] Support Every model provider in Agent Json feature
# Overview

Add support for all SDK model providers in the experimental `agent_config.py` (JSON-based agent configuration) feature. Currently, the `model` field only accepts a simple string which is always interpreted as a Bedrock `model_id`. This change will allow users to configure any of the 12 SDK model providers from JSON, enabling use cases like agent-builder tools and the `use_agent` tool to work with any provider.

## Implementation Requirements

Based on discussion and repository analysis:

### 1. Extend JSON Schema — Dual-Format `model` Field

The `model` field in `AGENT_CONFIG_SCHEMA` must support **both** formats for backward compatibility:

**String format (existing, backward compat — defaults to Bedrock):**
```json
{"model": "us.anthropic.claude-sonnet-4-20250514-v1:0"}
```

**Object format (new — supports all providers):**
```json
{
  "model": {
    "provider": "anthropic",
    "model_id": "claude-sonnet-4-20250514",
    "max_tokens": 10000,
    "client_args": {"api_key": "$ANTHROPIC_API_KEY"},
    "params": {"temperature": 0.7}
  }
}
```

Use a JSON Schema `oneOf` (or equivalent) to validate that `model` is either a string or an object with a required `provider` field plus provider-specific optional fields. The object format should use `"additionalProperties": true` (or a permissive schema) since each provider has different configuration parameters.

### 2. Environment Variable Resolution

Implement a utility function that recursively resolves environment variable references in config values. String values matching the pattern `$VAR_NAME` or `${VAR_NAME}` should be replaced with the corresponding environment variable value.

This allows secure configuration without embedding secrets:
```json
{
  "model": {
    "provider": "openai",
    "model_id": "gpt-4o",
    "client_args": {"api_key": "$OPENAI_API_KEY"}
  }
}
```

The resolution should be recursive (handle nested dicts and lists) and applied to the entire model config before provider instantiation.

### 3. Provider-to-Model Mapping

Implement a provider factory function in `agent_config.py` that maps provider name strings to their corresponding model class instantiation. Use the [strands-agents/tools model utilities](https://github.com/strands-agents/tools/tree/main/src/strands_tools/utils/models) as a reference pattern.

**All 12 SDK providers must be supported:**

| Provider Key | SDK Class | Module | Constructor Signature Notes |
|---|---|---|---|
| `bedrock` | `BedrockModel` | `strands.models.bedrock` | Supports `region_name`, `endpoint_url`, `boto_client_config` (dict→BotocoreConfig conversion needed) + `**model_config` |
| `anthropic` | `AnthropicModel` | `strands.models.anthropic` | `client_args` + `**model_config` (model_id, max_tokens, params) |
| `openai` | `OpenAIModel` | `strands.models.openai` | `client_args` + `**model_config` (model_id, params) |
| `gemini` | `GeminiModel` | `strands.models.gemini` | `client_args` + `**model_config` (model_id, params) |
| `ollama` | `OllamaModel` | `strands.models.ollama` | `host`, `ollama_client_args` + `**model_config` (model_id, additional_args, keep_alive, max_tokens, options, stop_sequences, temperature, top_p) |
| `litellm` | `LiteLLMModel` | `strands.models.litellm` | `client_args` + `**model_config` (model_id, params) |
| `mistral` | `MistralModel` | `strands.models.mistral` | `api_key`, `client_args` + `**model_config` (model_id, max_tokens, temperature, top_p, stream) |
| `llamaapi` | `LlamaAPIModel` | `strands.models.llamaapi` | `client_args` + `**model_config` (model_id, repetition_penalty, temperature, top_p, max_completion_tokens, top_k) |
| `llamacpp` | `LlamaCppModel` | `strands.models.llamacpp` | `base_url`, `timeout` + `**model_config` (model_id, params) |
| `sagemaker` | `SageMakerAIModel` | `strands.models.sagemaker` | `endpoint_config`, `payload_config`, `boto_client_config` (dict→BotocoreConfig) |
| `writer` | `WriterModel` | `strands.models.writer` | `client_args` + `**model_config` (model_id, max_tokens, stop, stream_options, temperature, top_p) |
| `openai_responses` | `OpenAIResponsesModel` | `strands.models.openai_responses` | `client_args` + `**model_config` (model_id, params, stateful) |

### 4. Provider-Specific Parameter Handling

Each provider has a different constructor signature. The factory function must correctly separate and route parameters:

- **Common pattern** (`client_args` + `**model_config`): anthropic, openai, gemini, litellm, llamaapi, writer, openai_responses
- **Bedrock special handling**: Extract `region_name`, `endpoint_url`, `boto_client_config` from the config dict. Convert `boto_client_config` from a plain dict to `botocore.config.Config` object.
- **Ollama**: Extract `host` and map `client_args` to `ollama_client_args` parameter name.
- **Mistral**: Extract `api_key` separately from `client_args`.
- **LlamaCpp**: Extract `base_url` and `timeout` from config.
- **SageMaker**: Extract `endpoint_config`, `payload_config`, `boto_client_config` as separate dict parameters. Convert `boto_client_config` from dict to `BotocoreConfig`.

### 5. Non-Serializable Parameters

The following constructor parameters **cannot** be specified from JSON (they require code-based instantiation) and should be documented as such:
- `boto_session` (BedrockModel, SageMakerAIModel) — use `region_name` instead
- `client` (OpenAIModel, GeminiModel) — use `client_args` instead
- `gemini_tools` (GeminiModel) — add programmatically after agent creation

### 6. Update `config_to_agent` Logic

In the `config_to_agent` function:
1. After parsing and validating the config, check if `model` is a string or dict
2. If string: pass directly to Agent constructor (existing behavior — becomes Bedrock model_id)
3. If dict: resolve environment variables, extract `provider`, call the factory function to create a `Model` instance, and pass the Model object to Agent constructor via `agent_kwargs["model"]`

### 7. Lazy Imports

All model provider imports must be lazy (at instantiation time, not module load time) to avoid requiring all optional dependencies. This matches the existing pattern in `src/strands/models/__init__.py`. Provide clear error messages if a provider's dependencies are not installed.

## Files to Modify

- `src/strands/experimental/agent_config.py` — Main implementation (schema, env var resolution, provider factory, updated `config_to_agent`)
- `tests/strands/experimental/test_agent_config.py` — Unit tests for all new functionality

## Test Requirements

- Test backward compatibility: string `model` field still works as Bedrock model_id
- Test object `model` field with each of the 12 providers (use mocking for provider classes)
- Test environment variable resolution: `$VAR`, `${VAR}`, nested dicts/lists, missing env vars
- Test schema validation: invalid provider, missing required `provider` field, invalid types
- Test special parameter handling: `boto_client_config` dict→BotocoreConfig, Ollama `host`, etc.
- Test error handling: unknown provider name, missing optional dependencies

## Acceptance Criteria

- [ ] `model` field accepts both string (Bedrock model_id) and object (provider + config) formats
- [ ] All 12 SDK model providers can be instantiated from JSON config
- [ ] Environment variable references (`$VAR` / `${VAR}`) are resolved in model config values
- [ ] Provider-specific constructor parameters are correctly routed
- [ ] Non-serializable parameters are handled gracefully (documented, clear errors)
- [ ] Existing tests continue to pass (backward compatibility)
- [ ] New unit tests cover each provider and edge cases
- [ ] Lazy imports prevent dependency issues for uninstalled optional providers
- [ ] JSON schema validates both string and object model formats
- [ ] `boto_client_config` dict-to-BotocoreConfig conversion works for Bedrock and SageMaker

## Assumptions

- No changes needed to the `Agent` constructor itself — the factory produces `Model` instances which the Agent already accepts
- No custom/virtual providers (e.g., "github", "cohere") — users can achieve this via `openai` provider with appropriate `client_args`
- No GitHub workflow changes needed
- This feature remains in `experimental` — the existing experimental disclaimer applies

## Notes for Future Improvement

- Consider adding a `model_providers` section to `AGENTS.md` documenting the JSON config format for each provider once this is stable
- The env var resolution utility could be reused for other config fields (e.g., tool configs) in the future
- Consider adding a JSON Schema `$ref` for each provider's config to enable IDE autocompletion

### PR DESCRIPTION
**Pull request by Unshure**
## Motivation

The experimental `agent_config.py` feature currently only accepts a simple string for the `model` field, which is always interpreted as a Bedrock `model_id`. This limits JSON-based agent configuration to a single provider, preventing use cases like agent-builder tools and the `use_agent` tool from working with any of the SDK's 12 model providers.

Resolves #1064

## Public API Changes

The `model` field in agent JSON configuration now supports two formats:

```python
# Before: string only (Bedrock model_id)
config = {"model": "us.anthropic.claude-sonnet-4-20250514-v1:0"}
agent = config_to_agent(config)

# After: string still works (backward compatible)
agent = config_to_agent(config)

# After: object format for any provider
config = {
    "model": {
        "provider": "openai",
        "model_id": "gpt-4o",
        "client_args": {"api_key": "..."}
    }
}
agent = config_to_agent(config)
```

All 12 SDK providers are supported: `bedrock`, `anthropic`, `openai`, `gemini`, `ollama`, `litellm`, `mistral`, `llamaapi`, `llamacpp`, `sagemaker`, `writer`, `openai_responses`. Each provider's constructor parameters are correctly routed — for example, `boto_client_config` dicts are converted to `BotocoreConfig` objects for Bedrock/SageMaker, Ollama's `client_args` maps to `ollama_client_args`, and Mistral's `api_key` is extracted as a separate parameter.

The factory logic is co-located with each provider via `from_dict` classmethods: the base `Model.from_dict` handles the common `client_args` + `**model_config` pattern (used by 7 providers), while Bedrock, Ollama, Mistral, LlamaCpp, and SageMaker override it for their non-standard constructors. `agent_config.py` is a thin orchestration layer that validates the schema, resolves the provider class via `strands.models` lazy `__getattr__`, and delegates to `from_dict`.

All provider imports are lazy to avoid requiring optional dependencies that aren't installed. Non-serializable parameters (`boto_session`, `client`, `gemini_tools`) cannot be specified from JSON and are documented as such.

## Use Cases

- **Agent-builder tools**: Create agents from JSON templates with any model provider, not just Bedrock
- **Multi-provider configs**: Store agent configurations in files that use OpenAI, Anthropic, or local models

### CODE REVIEW

File: src/strands/experimental/agent_config.py [hunk 1]
```diff
@@ -9,15 +9,44 @@
...
+    {
+        "model": {
+            "provider": "anthropic",
+            "model_id": "claude-sonnet-4-20250514",
+            "max_tokens": 10000,
+            "client_args": {"api_key": "$ANTHROPIC_API_KEY"}
+        }
+    }
+
+Environment variable references (``$VAR`` or ``${VAR}``) in model config values are resolved
```
Comment (Unshure) [inline src/strands/experimental/agent_config.py]: Let's remove the env var resolution for now

File: src/strands/experimental/agent_config.py [hunk 2]
```diff
@@ -50,6 +94,350 @@
 # Pre-compile validator for better performance
 _VALIDATOR = jsonschema.Draft7Validator(AGENT_CONFIG_SCHEMA)
 
+# Pattern for matching environment variable references
+_ENV_VAR_PATTERN = re.compile(r"^\$\{([^}]+)\}$|^\$([A-Za-z_][A-Za-z0-9_]*)$")
```
[the remark this answers is not in the document]
Comment (Unshure) [reply src/strands/experimental/agent_config.py]: Let's just remove env var for now

File: tests/strands/experimental/test_agent_config.py [hunk 3]
```diff
@@ -170,3 +179,292 @@ def test_config_to_agent_with_tool():
...
+        "litellm",
+        "mistral",
+        "llamaapi",
+        "llamacpp",
+        "sagemaker",
+        "writer",
+        "openai_responses",
+    ]
+
+    def test_all_providers_registered(self):
```
Comment (Unshure) [inline tests/strands/experimental/test_agent_config.py]: Let's remove this test, it doesn't add much value

File: tests/strands/experimental/test_agent_config.py [hunk 4]
```diff
@@ -170,3 +179,292 @@ def test_config_to_agent_with_tool():
...
+                    "provider": "openai",
+                    "model_id": "gpt-4o",
+                    "client_args": {"api_key": "test"},
+                    "custom_field": "allowed",
+                }
+            }
+            # Should not raise
+            config_to_agent(config)
+
+    def test_null_model_still_valid(self):
```
Comment (Unshure) [inline tests/strands/experimental/test_agent_config.py:239]: Is this already tested elsewhere? Can you also check and make sure there aren't duplicated tests?

**Comment by yonib05** [comment]
Closing as stale; the integration-test workflow has been substantially reworked on main (renamed to python-integration-test.yml). Branch will be deleted.
